### PR TITLE
[modelio] Update for iOS 10 beta 1

### PR DESCRIPTION
**Pull request by VincentDondain**
Notes:

- `initBoxWithExtent:segments:inwardNormals:geometryType:allocator:` == `newBoxWithDimensions:segments:geometryType:inwardNormals:allocator:`
- `initCylinderWithExtent:segments:inwardNormals:topCap:bottomCap:geometryType:allocator:` == `newCylinderWithHeight:radii:radialSegments:verticalSegments:geometryType:inwardNormals:allocator:`
- `initPlaneWithExtent:segments:geometryType:allocator:` == `newPlaneWithDimensions:segments:geometryType:allocator:`
- `initIcosahedronWithExtent:inwardNormals:geometryType:allocator:` == `newIcosahedronWithRadius:inwardNormals:allocator:`

Changed MDLVoxelArray base type to MDLObject (available on all platforms).

radar://26939747 fixed by Apple. https://trello.com/c/6aIzLH4a updated accordingly. MDLMaterialProperty "Copy" test re-enabled and it now passes.

Misses monotouch-tests [incoming].

### CODE REVIEW

**Comment by xamarin-release-manager** [comment]
Build failure

**Comment by VincentDondain** [comment]
4 tests failed **[In Progress]**

- introspection-mac (Unified, Unified XM45)
- link all (iOS Unified 64-bits)
- monotouch-test (watchOS) – _I think I can ignore this one._

File: src/ModelIO/MDLMesh.cs [hunk 1]
```diff
@@ -26,6 +27,53 @@
...
+				InitializeHandle (InitSphere (extent, segments, inwardNormals, geometryType, allocator), "initSphereWithExtent:segments:inwardNormals:geometryType:allocator:");
+			}
+		}
+
+		internal MDLMesh (MDLMesh mesh, int submeshIndex, uint subdivisionLevels, IMDLMeshBufferAllocator allocator)
+		{
+			InitializeHandle (InitMesh (mesh, submeshIndex, subdivisionLevels, allocator), "initMeshBySubdividingMesh:submeshIndex:subdivisionLevels:allocator:");
+		}
+
+		public MDLMesh CreateSphere (Vector3 dimensions, Vector2i segments, MDLGeometryType geometryType, bool inwardNormals, IMDLMeshBufferAllocator allocator)
```
Comment (spouliot) [inline src/ModelIO/MDLMesh.cs]: there's no usage of `this` and you're returning a new instance, any reason why this is **not** `static` ?

File: src/ModelIO/MDLMesh.cs [hunk 2]
```diff
@@ -26,6 +27,53 @@
...
+		{
+			return new MDLMesh (dimensions, segments, inwardNormals, geometryType, allocator, null, null, null);
+		}
+
+		public MDLMesh CreateHemisphere (Vector3 dimensions, Vector2i segments, MDLGeometryType geometryType, bool inwardNormals, bool cap, IMDLMeshBufferAllocator allocator)
```
Comment (spouliot) [inline src/ModelIO/MDLMesh.cs]: same

File: src/ModelIO/MDLMesh.cs [hunk 3]
```diff
@@ -26,6 +27,53 @@
...
+		{
+			return new MDLMesh (dimensions, segments, inwardNormals, geometryType, allocator, null, cap, false);
+		}
+
+		public MDLMesh CreateCapsule (Vector3 dimensions, Vector2i segments, MDLGeometryType geometryType, bool inwardNormals, int hemisphereSegments, IMDLMeshBufferAllocator allocator)
```
Comment (spouliot) [inline src/ModelIO/MDLMesh.cs]: same

File: src/ModelIO/MDLMesh.cs [hunk 4]
```diff
@@ -26,6 +27,53 @@
...
+		{
+			return new MDLMesh (dimensions, segments, inwardNormals, geometryType, allocator, hemisphereSegments, null, null);
+		}
+
+		public MDLMesh CreateCone (Vector3 dimensions, Vector2i segments, MDLGeometryType geometryType, bool inwardNormals, bool cap, IMDLMeshBufferAllocator allocator)
```
Comment (spouliot) [inline src/ModelIO/MDLMesh.cs]: same

File: src/ModelIO/MDLMesh.cs [hunk 5]
```diff
@@ -26,6 +27,53 @@
...
+		{
+			return new MDLMesh (dimensions, segments, inwardNormals, geometryType, allocator, null, cap, true);
+		}
+
+		public MDLMesh CreateSubdividedMesh (MDLMesh mesh, int submeshIndex, uint subdivisionLevels, IMDLMeshBufferAllocator allocator)
```
Comment (spouliot) [inline src/ModelIO/MDLMesh.cs]: same

File: src/ModelIO/MDLMesh.cs (hunk 1 above)
Comment (spouliot) [inline src/ModelIO/MDLMesh.cs]: Also if it's new and public then it likely needs some availability attributes

File: src/modelio.cs [hunk 6]
```diff
@@ -67,6 +67,9 @@ interface MDLAsset : NSCopying
 		[Export ("initWithURL:vertexDescriptor:bufferAllocator:")]
 		IntPtr Constructor (NSUrl url, [NullAllowed] MDLVertexDescriptor vertexDescriptor, [NullAllowed] IMDLMeshBufferAllocator bufferAllocator);
 
+		[Export ("initWithBufferAllocator:")]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 7]
```diff
@@ -82,6 +85,9 @@ interface MDLAsset : NSCopying
 		[Export ("canExportFileExtension:")]
 		bool CanExportFileExtension (string extension);
 
+		[Export ("childObjectsOfClass:")]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 8]
```diff
@@ -134,6 +140,22 @@ interface MDLAsset : NSCopying
 		MDLAsset FromScene (SCNScene scene, [NullAllowed] IMDLMeshBufferAllocator bufferAllocator);
 	}
 
+	[Protocol, Model]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 9]
```diff
@@ -143,6 +165,9 @@ interface MDLCamera
 			[MarshalDirective (NativePrefix = "xamarin_simd__", Library = "__Internal")] get;
 		}
 
+		[Export ("projection", ArgumentSemantic.Assign)]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 10]
```diff
@@ -387,6 +412,9 @@ interface MDLMaterial : MDLNamed, INSFastEnumeration
 		[Export ("count")]
 		nuint Count { get; }
 
+		[Export ("materialFace", ArgumentSemantic.Assign)]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 11]
```diff
@@ -484,12 +509,69 @@ interface MDLMaterialProperty : MDLNamed
 			[MarshalDirective (NativePrefix = "xamarin_simd__", Library = "__Internal")] get;
 			[MarshalDirective (NativePrefix = "xamarin_simd__", Library = "__Internal")] set;
 		}
+
+		[Export ("luminance")]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 12]
```diff
@@ -484,12 +509,69 @@ interface MDLMaterialProperty : MDLNamed
...
+
+		[Export ("connections")]
+		MDLMaterialPropertyConnection[] Connections { get; }
 	}
 
 	[iOS (9,0), Mac(10,11, onlyOn64 : true)]
 	[BaseType (typeof(MDLObject))]
 	interface MDLMesh
 	{
+		[Export ("initWithBufferAllocator:")]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 13]
```diff
@@ -501,6 +583,10 @@ interface MDLMesh
 		[return: NullAllowed]
 		MDLVertexAttributeData GetVertexAttributeDataForAttribute (string attributeName);
 
+		[Export ("vertexAttributeDataForAttributeNamed:asFormat:")]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 14]
```diff
@@ -510,18 +596,29 @@ interface MDLMesh
...
 		[Export ("vertexBuffers", ArgumentSemantic.Retain)]
 		IMDLMeshBuffer[] VertexBuffers { get; }
 
-		[Export ("submeshes", ArgumentSemantic.Retain)]
-		NSMutableArray<MDLSubmesh> Submeshes { get; }
+		[NullAllowed]
+		[Export ("submeshes", ArgumentSemantic.Copy)]
+		NSMutableArray<MDLSubmesh> Submeshes { get; set; }
+
+		[Export ("allocator", ArgumentSemantic.Retain)]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 15]
```diff
@@ -1286,13 +1438,21 @@ interface MDLTransformComponent : MDLComponent
 		}
 
 		[Abstract]
+		[Export ("resetsTransform")]
```
Comment (spouliot) [inline src/modelio.cs]: you cannot add new `[Abstract]` members, it's a breaking change

File: src/modelio.cs [hunk 16]
```diff
@@ -1286,13 +1438,21 @@ interface MDLTransformComponent : MDLComponent
...
+
+		[Abstract]
 		[Export ("minimumTime")]
 		double MinimumTime { get; }
 
 		[Abstract]
 		[Export ("maximumTime")]
 		double MaximumTime { get; }
 
+		[Abstract]
```
Comment (spouliot) [inline src/modelio.cs]: same

File: src/modelio.cs [hunk 17]
```diff
@@ -1438,17 +1601,24 @@ interface MDLVertexDescriptor : NSCopying
...
+		[Obsoleted (PlatformName.iOS, 10, 0)]
 		[Export ("initWithAsset:divisions:interiorShells:exteriorShells:patchRadius:")]
 		IntPtr Constructor (MDLAsset asset, int divisions, int interiorShells, int exteriorShells, float patchRadius);
 
+		[Deprecated (PlatformName.MacOSX, 10, 12)]
+		[Obsoleted (PlatformName.iOS, 10, 0)]
 		[Export ("initWithAsset:divisions:interiorNBWidth:exteriorNBWidth:patchRadius:")]
 		IntPtr Constructor (MDLAsset asset, int divisions, float interiorNBWidth, float exteriorNBWidth, float patchRadius);
 
+		[Export ("initWithAsset:divisions:patchRadius:")]
```
Comment (spouliot) [inline src/modelio.cs]: availability attributes ?

File: src/modelio.cs [hunk 18]
```diff
@@ -1438,17 +1601,24 @@ interface MDLVertexDescriptor : NSCopying
 	}
 
 	[iOS (9,0),Mac(10,11, onlyOn64 : true)]
-	[BaseType (typeof(NSObject))]
+	[BaseType (typeof(MDLObject))]
 	[DisableDefaultCtor]
 	interface MDLVoxelArray
 	{
 
+		[Deprecated (PlatformName.MacOSX, 10, 12)]
```
Comment (spouliot) [inline src/modelio.cs]: add `message` to suggest alternative

**Comment by spouliot** [comment]
I did not comment all cases, but every new API in existing types requires it's availability attributes.

File: src/ModelIO/MDLMesh.cs (hunk 1 above)
Comment (VincentDondain) [inline src/ModelIO/MDLMesh.cs]: Yes I just forgot the static keyword (which was easy to spot with the tests)

File: src/modelio.cs [hunk 19]
```diff
@@ -1601,21 +1661,29 @@ interface MDLVertexDescriptor : NSCopying
 	}
 
 	[iOS (9,0),Mac(10,11, onlyOn64 : true)]
-	[BaseType (typeof(MDLObject))]
```
Comment (spouliot) [inline src/modelio.cs]: sure ? because that's a breaking change on the Mac

**Comment by spouliot** [comment]
> Unified (Failed) Show details
> Unified XM45 (Failed) Show details

intro-mac needs to be taught about 

    [FAIL] Selector not found for ModelIO.MDLMaterialProperty : copyWithZone:

**Comment by spouliot** [comment]
@VincentDondain is the title accurate, i.e. up to with beta does that PR provide ?

If not 6, do you prefer completing it before merging ? or in another PR ?

**Comment by VincentDondain** [comment]
The work for following betas is in progress, now that this one is green I'd say we merge and I do a new PR (:

**Pull request merged**